modmail/config.py

```diff
@@ -1,144 +1,740 @@
-import asyncio
-import datetime
-import json
+import functools
+import inspect
 import logging
 import os
-import sys
+import pathlib
+import types
 import typing
-from pathlib import Path
-from typing import Any, Dict, Optional, Tuple
+from collections import defaultdict
 
+import attr
+import desert
 import discord
-import toml
-from discord.ext.commands import BadArgument
-from pydantic import BaseModel
-from pydantic import BaseSettings as PydanticBaseSettings
-from pydantic import Field, SecretStr
-from pydantic.color import Color as ColorBase
-from pydantic.env_settings import SettingsSourceCallable
-from pydantic.types import conint
-
-
-log = logging.getLogger(__name__)
-
-CONFIG_PATHS: list = [
-    f"{os.getcwd()}/config.toml",
-    f"{os.getcwd()}/modmail/config.toml",
-    "./config.toml",
+import discord.ext.commands
+import discord.ext.commands.converter
+import marshmallow
+import marshmallow.fields
+import marshmallow.validate
+
+from modmail.log import ModmailLogger
+
+
+_log_optional_deps = logging.getLogger("modmail.optional_dependencies")
+_silence = " You can silence these alerts by ignoring the `modmail.optional_dependencies` logger"
+try:
+    import dotenv
+except ModuleNotFoundError:  # pragma: nocover
+    _log_optional_deps.notice("dotenv was unable to be imported." + _silence)
+    dotenv = None
+
+try:
+    import atoml
+except ModuleNotFoundError:  # pragma: nocover
+    _log_optional_deps.notice("atoml was unable to be imported." + _silence)
+    atoml = None
+try:
+    import yaml
+except ModuleNotFoundError:  # pragma: nocover
+    _log_optional_deps.notice("yaml was unable to be imported." + _silence)
+    yaml = None
+
+del _log_optional_deps
+del _silence
+
+__all__ = [
+    "AUTO_GEN_FILE_NAME",
+    "CONFIG_DIRECTORY",
+    "ENV_PREFIX",
+    "USER_CONFIG_FILE_NAME",
+    "USER_CONFIG_FILES",
+    "ConfigLoadError",
+    "Config",
+    "config",
+    "ConfigurationSchema",
+    "BotConfig",
+    "BotModeConfig",
+    "ColourConfig",
+    "DeveloperConfig",
+    "convert_to_color",
+    "get_config",
+    "get_default_config",
+    "load_env",
+    "load_toml",
+    "load_yaml",
 ]
 
-DEFAULT_CONFIG_PATHS = [os.path.join(os.path.dirname(__file__), "config-default.toml")]
 
+logger: ModmailLogger = logging.getLogger(__name__)
 
-def determine_file_path(
-    paths=typing.Union[list, tuple], config_type: str = "default"
-) -> typing.Union[str, None]:
-    path = None
-    for file_path in paths:
-        config_file = Path(file_path)
-        if (config_file).exists():
-            path = config_file
-            log.debug(f"Found {config_type} config at {file_path}")
-            break
-    return path or None
 
+@functools.lru_cache
+def _get_config_directory() -> pathlib.Path:
```

Comment:
Might add it in a seperate pr, I don't want to add this here since it would involve adding a cli to the bot's entrypoint. 

While I'm aware that's not hard, I don't want to reach much further out of the scope of this pr.